spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/listener/AbstractHazelcastEventListener.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.hazelcast.listener;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.EventObject;
+import java.util.HashSet;
+import java.util.Set;
+
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.MultiMap;
+
+import org.springframework.integration.hazelcast.common.CacheListeningPolicyType;
+import org.springframework.integration.hazelcast.context.ApplicationContextStartEventHandler;
+import org.springframework.integration.hazelcast.inbound.HazelcastMessageProducer;
+import org.springframework.messaging.support.MessageBuilder;
+
+/**
+ * This is abstract Hazelcast event listener for entry/item/message events.
+ * 
+ * @author Eren Avsarogullari
+ * @since 1.0.0
+ *
+ */
+public abstract class AbstractHazelcastEventListener {
+
+	private HazelcastMessageProducer hazelcastInboundChannelAdapter;
+
+	public AbstractHazelcastEventListener(HazelcastMessageProducer hazelcastInboundChannelAdapter) {
+		this.hazelcastInboundChannelAdapter = hazelcastInboundChannelAdapter;
+	}
+
+	public HazelcastMessageProducer getHazelcastInboundChannelAdapter() {
+		return hazelcastInboundChannelAdapter;
+	}
+
+	protected abstract void processEvent(EventObject event);
+
+	protected Set<SocketAddress> getLocalSocketAddresses() {
+		Set<SocketAddress> localSocketAddressesSet = new HashSet<>();
+		for (HazelcastInstance hazelcastInstance : Hazelcast.getAllHazelcastInstances()) {
+			localSocketAddressesSet.add(hazelcastInstance.getLocalEndpoint().getSocketAddress());
+		}
+
+		return localSocketAddressesSet;
+	}
+
+	protected void sendMessage(EventObject event, InetSocketAddress socketAddress,
+			CacheListeningPolicyType cacheListeningPolicyType) {
+		
+		if (CacheListeningPolicyType.ALL == cacheListeningPolicyType
+				|| isEventAcceptable(socketAddress)) {
+
+			getHazelcastInboundChannelAdapter().sendMessage(
+					MessageBuilder.withPayload(event).build());
```

Comment:
I agree about this approach. I think the following design is ok.

HazelcastHeaders covers **event, member, name, publishingTime.**

For EntryEvent :
Header : event, member and name.
Payload : key, oldValue and value(**EntryEventMessagePayload** object can be created and covers these fields)

For ItemEvent : 
Header : event and member.
Payload : item

For MessageEvent : 
Header : member, publishingTime, name
Payload : message